File: ruby/lib/dsa/singly_linked_list.rb
module DSA
  # A singly linked list.
  #
  # Each item in the list contains a value and an optional pointer to the next
  # item in the list.
  #
  class SinglyLinkedList
    class Node
      # @return [T]
      attr_accessor :value

      # @return [Node, nil]
      attr_accessor :next

      # @param value [T]
      # @param next [Node, nil]
      #
      def initialize value:, next: nil
        @value = value
        @next = binding.local_variable_get(:next)
      end
    end

    # Add an item to the front of the list.
    #
    # - Time: O(1), since we don't have to iterate the list
    # - Space: O(1), no additional space based on input size
    #
    # @param value [T]
    # @return [Node] the new front node
    #
    def prepend value
      @root = Node.new(value:, next: @root)
    end

    # Yield each item of the list.
    #
    # - Time: O(n), since we iterate the list
    # - Space: O(1), no additional space based on input size
    #
    # @yield [Node] each item of the list, if a block is given
    # @return [Enumerator<Node>] if no block is given
    #
    def each
      return to_enum(__method__) unless block_given?

      return if empty?

      curr = @root

      while curr.next
        yield curr
        curr = curr.next
      end

      yield curr
    end

    # Count how many items are in the list
    #
    # - Time: O(n), since we have to iterate the list and count the items
    # - Space: O(1), no additional space based on input size
    #
    # @return [Integer] the list length
    #
    def length = each.count

    # Get the node at the given index
    #
    # - Time: O(n), since we have to iterate the list to find the node at that index
    # - Space: O(1), no additional space based on input size
    #
    # @return [Node]
    #
    def [] index
      raise ArgumentError, "index must be an integer" unless index.respond_to?(:to_i) && index.to_i == index
      raise ArgumentError, "index must be 0 or greater" unless index >= 0
      raise ArgumentError, "list is empty" if empty?

      each.with_index do |item, i|
        return item if i == index
      end

      raise ArgumentError, "index is too large"
    end

    # Delete an item
    #
    # - Time: O(n), since we have to iterate the list to find the previous node
    # - Space: O(1), no additional space based on input size
    #
    # @param node [Node] the node to delete
    # @return [Node] the deleted node
    #
    def delete node:
      raise ArgumentError, "node must be a node" unless node.is_a?(Node)

      if @root == node
        @root = node.next
        return node
      end

      each do |item|
        if item.next == node
          item.next = node.next
          return node
        end
      end

      raise ArgumentError, "node isn't in the list"
    end

    # Insert an item at the given index.
    #
    # - Time: O(n), we have to iterate the list to find the previous index's node
    # - Space: O(1), no additional space based on input size
    #
    # @param value [T]
    # @param index [Integer]
    # @return [Node] the new node
    #
    def insert value:, index:
      raise ArgumentError, "index must be an integer" unless index.respond_to?(:to_i) && index.to_i == index
      raise ArgumentError, "index must be 0 or greater" unless index >= 0

      return @root = Node.new(value:, next: @root) if index.zero?

      each.with_index do |item, i|
        return item.next = Node.new(value:, next: item.next) if i + 1 == index
      end

      raise ArgumentError, "index is too large"
    end

    # Append an item to the end of the list.
    #
    # - Time: O(n), since we have to find the last node
    # - Space: O(1), no additional space based on input size
    #
    # @param value [T]
    # @return [Node] the new last node
    #
    def append value
      return @root = Node.new(value:) if empty?

      last = @root
      each { |item| last = item }
      last.next = Node.new(value:)
    end

    # Check if an item exists in the list.
    #
    # Time: O(n), since we have to find the first node containing the item
    # Space: O(1), no additional space based on input size
    #
    # @param value [T]
    # @return [Boolean] if the item exists in the list
    #
    def include? value
      each { |item| return true if item.value == value }
      false
    end

    # Check if the list is empty.
    #
    # - Time: O(1), since we don't have to iterate the list
    # - Space: O(1), no additional space based on input size
    #
    # @return [Boolean] whether the list is empty
    #
    def empty? = @root.nil?

    # Return a string representation of the list
    #
    # @return [String]
    #
    def to_s = each.map(&:value).join(" -> ")
  end
end

File: zig/src/singly_linked_list.zig
const std = @import("std");
const testing = std.testing;

fn Node(comptime T: type) type {
    return struct {
        const Self = @This();

        value: T,
        next: ?*Self,
    };
}

/// A singly linked list.
///
/// Each item in the list contains a value and an optional pointer to the next
/// item in the list.
///
pub fn SinglyLinkedList(comptime T: type) type {
    return struct {
        const Self = @This();

        allocator: std.mem.Allocator,
        root: ?*Node(T),

        /// Create a new list.
        ///
        pub fn init(allocator: std.mem.Allocator) Self {
            return Self{
                .root = null,
                .allocator = allocator,
            };
        }

        /// Release all allocated memory.
        ///
        pub fn deinit(self: *Self) void {
            if (self.root) |root| {
                var node = root;

                while (node.next) |next| {
                    self.allocator.destroy(node);
                    node = next;
                }
                self.allocator.destroy(node);
            }

            self.* = undefined;
        }

        /// Add an item to the beginning of the list.
        ///
        pub fn prepend(self: *Self, item: T) !*Node(T) {
            const node = try self.allocator.create(Node(T));
            node.* = Node(T){ .value = item, .next = self.root };

            self.root = node;

            return node;
        }

        /// Count how many items are in the list.
        ///
        pub fn length(self: *Self) usize {
            var sum: usize = 0;

            if (self.root) |root| {
                var current = root;
                sum += 1;

                while (current.next) |next| {
                    current = next;
                    sum += 1;
                }
            }

            return sum;
        }

        /// Append an item to the end of the list.
        ///
        pub fn append(self: *Self, item: T) !*Node(T) {
            if (self.root) |root| {
                var current = root;
                while (current.next) |next| {
                    current = next;
                }
                const last = try self.allocator.create(Node(T));
                last.* = Node(T){ .value = item, .next = null };

                current.next = last;
                return last;
            }

            const node = try self.allocator.create(Node(T));
            node.* = Node(T){ .value = item, .next = null };

            self.root = node;
            return node;
        }

        /// Get the node at the given index.
        ///
        pub fn get(self: *Self, index: usize) !*Node(T) {
            if (self.root) |root| {
                var current = root;

                if (index == 0) {
                    return current;
                }

                var current_index: usize = 1;

                while (current.next) |next| {
                    if (current_index == index) {
                        return next;
                    }
                    current_index += 1;
                    current = next;
                }

                if (current_index == index + 1) {
                    return current;
                }

                return error.IndexOutOfBounds;
            }

            return error.EmptyList;
        }

        /// Delete a node.
        ///
        pub fn delete(self: *Self, node: *Node(T)) !void {
            if (self.root) |root| {
                if (root == node) {
                    self.root = root.next;
                    self.allocator.destroy(root);
                    return;
                }

                var current = root;

                while (current.next) |next| {
                    if (next == node) {
                        current.next = next.next;
                        self.allocator.destroy(next);
                        return;
                    }
                    current = next;
                }
            }

            return error.NotInList;
        }

        /// Insert an item at the given index.
        ///
        pub fn insert(self: *Self, item: T, index: usize) !*Node(T) {
            if (self.root) |root| {
                if (index == 0) {
                    return self.prepend(item);
                }

                var current = root;
                var current_index: usize = 0;

                while (current.next) |next| {
                    if (current_index + 1 == index) {
                        const node = try self.allocator.create(Node(T));
                        node.* = Node(T){ .value = item, .next = next };

                        current.next = node;

                        return node;
                    }
                    current_index += 1;
                    current = next;
                }

                if (current_index + 1 == index) {
                    const node = try self.allocator.create(Node(T));
                    node.* = Node(T){ .value = item, .next = null };

                    current.next = node;

                    return node;
                }

                return error.IndexOutOfBounds;
            }

            if (index == 0) {
                return self.prepend(item);
            }

            return error.IndexOutOfBounds;
        }

        /// Check if an item exists in the list.
        ///
        pub fn include(self: *Self, item: T) !bool {
            if (self.root) |root| {
                var current = root;
                while (current.next) |next| {
                    if (current.value == item) {
                        return true;
                    }

                    current = next;
                }
                if (current.value == item) {
                    return true;
                }
            }

            return false;
        }

        /// Check if the list is empty.
        ///
        pub fn is_empty(self: *Self) bool {
            return self.root == null;
        }
    };
}

test {
    testing.refAllDecls(@This());
}

test "init/deinit" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();
}

test "prepend" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    const node = try list.prepend(42);
    try testing.expectEqual(@as(i32, 42), node.value);
}

test "append" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    const node = try list.append(42);
    try testing.expectEqual(@as(i32, 42), node.value);
}

test "length" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    try testing.expectEqual(@as(usize, 0), list.length());

    _ = try list.prepend(42);
    try testing.expectEqual(@as(usize, 1), list.length());

    _ = try list.prepend(42);
    try testing.expectEqual(@as(usize, 2), list.length());
}

test "get" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    try testing.expectError(error.EmptyList, list.get(0));

    var node: *Node(i32) = undefined;

    _ = try list.prepend(2);
    _ = try list.prepend(1);
    _ = try list.append(3);

    node = try list.get(0);
    try testing.expectEqual(@as(i32, 1), node.value);

    node = try list.get(1);
    try testing.expectEqual(@as(i32, 2), node.value);

    node = try list.get(2);
    try testing.expectEqual(@as(i32, 3), node.value);

    try testing.expectError(error.IndexOutOfBounds, list.get(3));
}

test "delete" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    _ = try list.append(1);
    _ = try list.append(2);
    _ = try list.append(3);

    var missing = Node(i32){ .next = null, .value = 42 };
    try testing.expectError(error.NotInList, list.delete(&missing));

    const one = try list.get(0);
    const two = try list.get(1);
    const three = try list.get(2);

    try testing.expectEqual(@as(usize, 3), list.length());

    try list.delete(three);
    try testing.expectEqual(@as(usize, 2), list.length());

    try list.delete(two);
    try testing.expectEqual(@as(usize, 1), list.length());

    try list.delete(one);
    try testing.expectEqual(@as(usize, 0), list.length());
}

test "insert" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    var node: *Node(i32) = undefined;

    node = try list.insert(1, 0);
    try testing.expectEqual(@as(i32, 1), node.value);

    node = try list.insert(2, 1);
    try testing.expectEqual(@as(i32, 2), node.value);

    node = try list.insert(3, 2);
    try testing.expectEqual(@as(i32, 3), node.value);

    const one = try list.get(0);
    const two = try list.get(1);
    const three = try list.get(2);

    try testing.expectEqual(@as(i32, 1), one.value);
    try testing.expectEqual(@as(i32, 2), two.value);
    try testing.expectEqual(@as(i32, 3), three.value);

    try testing.expectError(error.IndexOutOfBounds, list.insert(42, 4));
}

test "include" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    _ = try list.append(1);
    _ = try list.append(2);
    _ = try list.append(3);

    try testing.expectEqual(true, try list.include(1));
    try testing.expectEqual(true, try list.include(2));
    try testing.expectEqual(true, try list.include(3));
    try testing.expectEqual(false, try list.include(42));
}

test "is_empty" {
    var list = SinglyLinkedList(i32).init(testing.allocator);
    defer list.deinit();

    try testing.expectEqual(true, list.is_empty());

    _ = try list.append(1);

    try testing.expectEqual(false, list.is_empty());
}
